### DISCUSSION TITLE
lazy='noload' strange behavior

### DISCUSSION THREAD

**Discussion by miguelgrinberg** [sqlalchemy/sqlalchemy#7377] [Usage Questions] [answered]
Hey, making more progress on upgrading some of my pre 1.4 projects to use 1.4 syntax, and I came upon a strange issue.

Using the one-to-many relationship between users and blog posts as an example. Even if the `User.posts` relationship is set to `lazy='noload'`, adding a post through this relationship works fine:

```python
>>> user.posts.append(new_post)
>>> session.commit()
```

But removing a post fails, because the same logic used with a populated relationship appears to be used:

```python
>>> user.posts.remove(new_post)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/Users/mgrinberg/Documents/venv/lib/python3.10/site-packages/sqlalchemy/orm/collections.py", line 1179, in remove
    fn(self, value)
ValueError: list.remove(x): x not in list
```

Would you agree that it is expected that when using `noload` the remove logic should be adjusted to work without a loaded relationship? Or is there a different/better way to remove without having to manually work with the foreign key directly?

**Comment by zzzeek** [comment, marked as answer]
the best outcome with "noload" would be that we can remove it from the ORM entirely, because it is nonsensical.  however, people seem to use it for random things here and there.     so being able to improve it, or remove it , would be based on the things people actually are using it for, what the problems are which they're trying to solve, and coming up with more airtight solutions.

that was kind of a digression though because in your specific case, I would say that it is not actually surprising that you can't remove items from a "noload" collection that you didn't actually place there directly, or after such collection had been expired.      you can remove items that you put there, which were also not expired, without issue:

```py
class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    bs = relationship("B", lazy="noload")


class B(Base):
    __tablename__ = "b"
    id = Column(Integer, primary_key=True)
    a_id = Column(ForeignKey("a.id"))


a1 = A()
b1 = B()

a1.bs.append(b1)
a1.bs.remove(b1)
```

but once it's expired, you've told it to "noload".  So, you can't expect that an object is in the collection when it isn't loaded.  I can see how this might be *inconvenient*, but we would assume "noload" is being used for some very specific reason that warrants the inconvenience of the colleciton not actually able to load things, yet still remaining available as a Python list.

**Reply by miguelgrinberg** [reply]
Hi Mike. Yes, I do understand the limitations of `noload`. It seems odd that I can add but not remove, though. 

> I can see how this might be inconvenient

I think it is more than an inconvenience. If I want to do this all in 2.x query style, then removing elements from a relationship via the ORM is not possible unless the collection is pre-loaded. But a lot of situations require collections that are not automatically loaded, so it seems in those cases this is not possible at all for async, and only possible for sync if you are willing to use legacy 1.x style constructs. Am I missing something in this analysis?

**Reply by miguelgrinberg** [reply]
Actually, it seems this is also not working with `lazy='dynamic'`. Weird, I expected that to work. If I try to remove from a dynamic collection I do not get any errors when I call `user.posts.remove(post)`. Checking the `post.user_id` attribute right after shows that it was reset to `None`. But a `session.commit()` does not update the post's `user_id` attribute.

Update: Hmm. Okay, this does work with dynamic, but only if the `post` object that was removed is explicitly added to the session.

**Reply by zzzeek** [reply]
> Hi Mike. Yes, I do understand the limitations of `noload`. It seems odd that I can add but not remove, though.

the list is empty, because it was expired.  so you can't remove anything from it.  if we made it silently ignore removals, then it would still be a bug if you said mylist.remove("random stuff"), because that wouldn't raise.   the list must be emptied on expiration because expiration means we no longer assume anything about the database, including that the objects which are inside of lists still exist in the database.

So im not sure what other alternatives there are.    If someone is using noload, they can't expire the collection.  you would set expire_on_commit=False on your Session.



> 
> > I can see how this might be inconvenient
> 
> I think it is more than an inconvenience. 

Again I don't think there are non-hacky reasons to use "noload".  When I've seen people use it, they are trying to implement some special hack.   Normal applications should not have to use "noload" for anything.


> If I want to do this all in 2.x query style, then removing elements from a relationship via the ORM is not possible unless the collection is pre-loaded. But a lot of situations require collections that are not automatically loaded, so it seems in those cases this is not possible at all for async, and only possible for sync if you are willing to use legacy 1.x style constructs. Am I missing something in this analysis?

I'm not sure what is leading to the use of "noload" here, unless this regards that I had previously mentioned it in conjunction with using the "with_parent()" function as a replacement for "dynamic".   as we discussed yesterday, I have removed references to "noload" for this use case.   2.x users who want to upgrade "dynamic" should for now keep the "dynamic" loader in place and make use of `session.scalars(user.posts.statement)` .   that will continue to work in 2.0 without any warnings.     later, there will be another "lazy" setting and they'd be able to use something more like `session.scalars(users.posts.select())`

**Reply by miguelgrinberg** [reply]
> So im not sure what other alternatives there are. If someone is using noload, they can't expire the collection. you would set expire_on_commit=False on your Session.

Just in case I wasn't completely clear before, what I'm trying to do is not just to remove an item that I added right before. I was wondering about a general case, where I want to unlink a random child from the parent while this relationship is not loaded. Without the ORM, this is as simple as setting the foreign key on the child to `None`, there is no need to have the whole collection loaded to do this. I expected doing a remove would work through the relationship in a similar way. The validation that the child is in the collection can be done via the foreign key, without access to the collection. I think?

> Again I don't think there are non-hacky reasons to use "noload".

Okay, this seems to be the misunderstanding on my part. I did not realize that the reference to using `noload` in the docs was now removed. That makes sense now.

So I guess there is currently no 2.x only way to say that you don't automatically want to load a relationship, while at the same time be able to use the relationship to add/remove. The only option right now is to use `dynamic` and live with the legacy query object. 

Will the new lazy mode described in https://github.com/sqlalchemy/sqlalchemy/issues/7123 be able to handle add/remove?

**Reply by CaselIT** [reply]
> Will the new lazy mode described in #7123 be able to handle add/remove?

I guess we could provide a `remove`/`add` methods so that you may be able to do `session.execute(User.posts.remove(the_post))` or `session.execute(User.posts.add(new_post))`. Not sure if an api like this would make sense?

**Reply by miguelgrinberg** [reply]
I was hoping it would work in a way that is compatible with the other lazy modes, which is in a collection-like way. I haven't thought about the many-to-many case which might be tricky, but for a one-to-many, adding and removing can be done by setting/resetting the foreign key on the "many" object, without having to have the collection in memory and without doing any I/O, so it would work well with async too.

**Reply by zzzeek** [reply]
> Just in case I wasn't completely clear before, what I'm trying to do is not just to remove an item that I added right before. I was wondering about a general case, where I want to unlink a random child from the parent while this relationship is not loaded. 

there is a way.  this is how lazy="dynamic" actually works, when you "add" or "remove" items, it doesn't mutate any collection, it just puts a memo internally to the effect of "add this B, remove that B" which takes place during flush.  I think, at least I hope, that during the flush if you were trying to remove something that wasn't there, it would raise an error, though im not sure I'd have to check what it does in that case.

as we are not getting rid of lazy="dynamic" and are instead going to make a new version of it which just doesn't iterate directly, but still supports add/remove the same way (it will use the same code), this use case remains available.

> Without the ORM, this is as simple as setting the foreign key on the child to `None`, there is no need to have the whole collection loaded to do this. I expected doing a remove would work through the relationship in a similar way. 

This is the same with the ORM too.  You go to the related object, and you can either a. set the actual FK attribute to None, or b. set the many-to-one side of the relationship to None, just as you describe.   This is as you know how it's always worked and it's exactly like if you were working with the tables directly.


> The validation that the child is in the collection can be done via the foreign key, without access to the collection. I think?

using a many-to-one relationship is the normal way people do this.

> 
> So I guess there is currently no 2.x only way to say that you don't automatically want to load a relationship, while at the same time be able to use the relationship to add/remove. The only option right now is to use `dynamic` and live with the legacy query object.

as well as the new version of "dynamic" that we will be adding, which will work the same way and use the same code.

> 
> Will the new lazy mode described in #7123 be able to handle add/remove?

yes, we are literally going to take the existing "dynamic" code and just make the ``__iter__()`` method raise, and say, please call ".select()".  it will likely be a flag on the existing object.

**Reply by miguelgrinberg** [reply]
Awesome. Thanks for the detailed response!
